ChiModules/KEigenvalueSolver/IterativeOperations/set_k_source.cc

```diff
@@ -43,95 +46,130 @@ void KEigenvalue::Solver::
-                else
-                  fission_g += xs->chi[g] * xs->nu_sigma_f[gprime] *
-                               phi_prevp[gprime] / k_eff;
+        //============================== Loop over groupset groups
+        for (int g = gs_i; g <= gs_f; ++g) {
+          //======================================== Apply scattering
+          double inscatter_g = 0.0;
+          if (ell < xs->transfer_matrices.size()) {
+            //============================== Across-groupset
+            if (apply_ags_scatter_src) {
+              size_t num_transfers =
+                  xs->transfer_matrices[ell].rowI_indices[g].size();
+
+              //============================== Loop over transfers
+              for (int t = 0; t < num_transfers; ++t) {
+                size_t gprime = xs->transfer_matrices[ell].rowI_indices[g][t];
+
+                if ((gprime < gs_i) or (gprime > gs_f)) {
+                  double sigma_sm = xs->transfer_matrices[ell].rowI_values[g][t];
+                  inscatter_g += sigma_sm * phi_oldp[gprime];
+                }
+              }
             }
-          }
-          q_mom[g] += fission_g;
-
-          // ----- Contribute precursors
-          double precursor_g = 0.0;
-          if ( (ell == 0) and (options.use_precursors) )
-          {
-            if ((apply_mat_src) and (xs->num_precursors > 0))
-            {
-              for (int j=0; j<num_precursors; ++j)
-              {
-                for (size_t gprime=first_grp; gprime<=last_grp; ++gprime)
-                {
-                  precursor_g += xs->chi_delayed[g][j] * xs->precursor_yield[j] *
-                                 xs->nu_delayed_sigma_f[gprime] *
-                                 phi_prevp[gprime] / k_eff;
+
+            //============================== Within-groupset
+            if (apply_wgs_scatter_src) {
+              size_t  num_transfers =
+                  xs->transfer_matrices[ell].rowI_indices[g].size();
+
+              //============================== Loop over transfers
+              for (int t = 0; t < num_transfers; ++t) {
+                size_t gprime = xs->transfer_matrices[ell].rowI_indices[g][t];
+
+                if ((gprime >= gs_i) and (gprime <= gs_f)) {
+                  double sigma_sm = xs->transfer_matrices[ell].rowI_values[g][t];
+                  inscatter_g += sigma_sm * phi_oldp[gprime];
                 }
               }
             }
-          }
-          q_mom[g] += precursor_g;
+          }//if moment avail
+
+          q_mom[g] += inscatter_g;
+
+          //======================================== Apply fission
+          if (xs->is_fissile and (ell == 0)) {
+            double fission_g = 0.0;
+            //============================== Across-groupset
+            if (apply_ags_fission_src) {
+              //============================== Loop over groups
+              for (size_t gprime = first_grp; gprime <= last_grp; ++gprime) {
```

Comment:
I am used to seeing C++ loops for var < var_max but here it is var <= var_max. Why <= here (and later too)?

Reply:
This is how it is done in the SetSource routine in the LinearBoltzmann::Solver class. This loops from the id of the first group in the list of group objects to the id of the last group in the list of group objects. Doing this does suggest that the first group has to be smallest id and the last group the largest. I am unsure why this was the decision made within the LinearBoltzmann::Solver class as opposed to looping from 0 -> num_groups. @Naktakala could probably answer this.